### DISCUSSION TITLE
Set process title for Windows

### DISCUSSION THREAD

**Discussion by igor-dan** [dvarrazzo/py-setproctitle#122] [Ideas]
I use following approach for Windows
You can set predefined process environment variable, for example
os.environ['PROCESS_NAME'] = 'myProcessName'

To check, if Python process with this name running from other Python process, you can use psutil package 
to iterate between all processes and check their environment variables.

**Comment by dvarrazzo** [comment]
So on windows everyone can see everyone else's env vars? Safe...

**Reply by gershnik** [reply]
It's no different from /proc/pid/environ on Linux where you can also see every process environment. Given sufficient permissions of course, which is also the same restriction as on Windows. Arguably Windows non elevated processes, which cannot do that, provide better protection.

**Reply by dvarrazzo** [reply]
But that means that only root (an Administrator, I guess, in Windows parlance) could use a tool to see the process title (as well as accidentally expose passwords etc). On Linux you need to be root in order to snoop into `/proc/PID`, whereas the customized title can be seen by everyone.

**Reply by gershnik** [reply]
Most likely. Historically psutil required admin rights. It might work for the processes of the same user these days though - I haven't touched Windows at that level in a long while. 
Personally I don't think there is a good solution on Windows unless Task Manager behavior changes. Then again if Task Manager were to show what an app asked it to, given proliferation of windows malware, it would likely be full of "important Microsoft program" entries 🙂🤷‍♂️

**Comment by ysc3839** [comment]
https://github.com/dvarrazzo/py-setproctitle/issues/40#issuecomment-2370190272
You could try modify the buffer returned by [`GetCommandLineA`](https://learn.microsoft.com/en-us/windows/win32/api/processenv/nf-processenv-getcommandlinea) and [`GetCommandLineW`](https://learn.microsoft.com/en-us/windows/win32/api/processenv/nf-processenv-getcommandlinew).
Doing so can't change process name, but can change process command line. And it's visible in Task Manager and Process Explorer.
